quinn-proto/src/connection/assembler.rs

```diff
@@ -7,19 +7,14 @@ use crate::range_set::RangeSet;
 /// Helper to assemble unordered stream frames into an ordered stream
 #[derive(Debug, Default)]
 pub(crate) struct Assembler {
-    /// Stream offset of the first byte unread by the application. Meaningless if `unordered` is
-    /// true.
-    offset: u64,
+    state: State,
     data: BinaryHeap<Chunk>,
     defragmented: usize,
-    /// Whether any unordered reads have been performed, making this assembler unusable for ordered
-    /// reads
-    unordered: bool,
     /// Number of bytes read. Equal to offset when `unordered` is false, may or may not be
     /// otherwise.
     bytes_read: u64,
-    /// Offsets received so far
-    recvd: RangeSet,
+    /// First offset we haven't received any data at or after
+    limit: u64,
```

Comment:
I feel like `limit` isn't a great name for this. Any other ideas? `end`, `next`, `unseen`? (Of these options I think I like `next` best.)

Reply:
I went with `end`, which had precedent elsewhere in related code. `next` could be confusing because it strictly has no bearing on what will actually be received next; data can be transmitted in any order so long as flow control limits are respected.